### PR TITLE
Migrate all rules to ESM

### PR DESCRIPTION
**Pull request by ybiquitous**
> Which issue, if any, is this issue related to?

Ref #5291

> Is there anything in the PR that needs further explanation?

I created and performed a temporary script to rewrite all rules, but several commits were needed for preparation. Please look at each commit to confirm my workflow, especially 131dc13b.

Note: The test coverage temporarily degrades, but it's due to `jest-preset-stylelint`. See <https://github.com/stylelint/stylelint/pull/7182#issuecomment-1743972736>.

---

The next step after this PR will be to remove `import-lazy` in `lib/rules/index.mjs` like #7184:

https://github.com/stylelint/stylelint/blob/8e66566ec9d65d3369e97163e5f38a1bd54caca2/lib/rules/index.mjs#L5

### CODE REVIEW

**Comment by ybiquitous** [comment]
FYI. I've just opened PR https://github.com/stylelint/jest-preset-stylelint/pull/77, which will enable us to test ESM rewritten rules.

**Review by mattxwang — COMMENTED** [review]
As a note, I've reviewed all the commits in the stack *other* than [131dc13](https://github.com/stylelint/stylelint/pull/7211/commits/131dc13b37c0855263e86763812404e7900bf577) so far, and they seem reasonable to me. I think I'll need to sit down for a while and make sure that I can properly review that one as well; however, being able to test rewritten rules would help with that significantly. Thanks for taking this on so far!

**Review by mattxwang — APPROVED** [review]
Sorry for the delay! I've sat down and taken as good of a look at [131dc13](https://github.com/stylelint/stylelint/pull/7211/commits/131dc13b37c0855263e86763812404e7900bf577) that I think I can, by:

- looking at `rewrite-rules-to-esm.mjs`, which seems correct
- manually taking a look at a few of the edited files to look for errors

With test coverage, I feel relatively confident that this is good to merge. However, it'd be great to have another set of eyes to approve this as well (perhaps @Mouvedia, who has also been pretty active recently)?

Appreciate you taking on this huge task @ybiquitous!

**Review by Mouvedia — COMMENTED** [review]
@mattxwang I tried.
@ybiquitous it seems you used a tool or script so I couldn't find much.

File: lib/rules/color-hex-alpha/index.mjs
Comment (ybiquitous) [reply lib/rules/color-hex-alpha/index.mjs]: Good catch. Fixed e62d3f851 👍🏼

**Comment by ybiquitous** [comment]
> it seems you used a tool or script so I couldn't find much.

@Mouvedia Please see `scripts/rewrite-rules-to-esm.mjs` (b2d3af5), which I removed with the commit since it was temporary.

File: lib/rules/color-hex-length/__tests__/index.mjs
Comment (ybiquitous) [reply lib/rules/color-hex-length/__tests__/index.mjs:2]: Yes. But I'll rename `.js` files in `fixtures/` with another PR to keep this PR's diff size.

File: lib/rules/color-hex-length/index.mjs
Comment (ybiquitous) [reply lib/rules/color-hex-length/index.mjs:139]: Good catch. I've opened #7245 for refactoring.

File: lib/rules/color-hex-alpha/index.mjs
Comment (Mouvedia) [reply lib/rules/color-hex-alpha/index.mjs]: I am surprised you care as well :)

Maybe we could add https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/order.md to https://github.com/stylelint/eslint-config-stylelint/blob/main/index.js?

Comment (ybiquitous) [reply lib/rules/color-hex-alpha/index.mjs]: Sounds good. Could you open an issue to discuss `eslint-plugin-import` on the `eslint-config-stylelint` repo?
https://github.com/stylelint/eslint-config-stylelint/issues/new/choose

We have used the built-in [`sort-imports`](https://eslint.org/docs/latest/rules/sort-imports) rule, but we should know pros/cons by comparing them:
https://github.com/stylelint/eslint-config-stylelint/blob/41df28e06546134480aeea91515959a59525273b/index.js#L110

Comment (Mouvedia) [reply lib/rules/color-hex-alpha/index.mjs]: stylelint/eslint-config-stylelint#241

**Comment by ybiquitous** [comment]
Since this PR has so many conflicts, I'll split this into several PRs.

**Comment by ybiquitous** [comment]
I'll merge extracted PRs when CI passes to avoid pain due to so many conflicts.

**Comment by ybiquitous** [comment]
This PR is no longer needed since PR #7262 has been merged.

**Pull request closed without merging**
